File: 01-Programming-Languages/JavaScript/Data-Structures/LinkedList/reverseLinkedList.js
/*
Problem: Reverse Linked List
Description: Reverse a singly linked list iteratively.

Approach:
- Use three pointers (prev, curr, next) to reverse links in one pass

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: 1 -> 2 -> 3 -> 4 -> 5
Output: 5 -> 4 -> 3 -> 2 -> 1
*/

class ListNode {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

function reverseList(head) {
  let prev = null, curr = head;
  while (curr) {
    const next = curr.next;
    curr.next = prev;
    prev = curr;
    curr = next;
  }
  return prev;
}

function printList(head) {
  const vals = [];
  while (head) { vals.push(head.val); head = head.next; }
  return vals.join(' -> ');
}

const list = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
console.log('Original:', printList(list));
console.log('Reversed:', printList(reverseList(list)));

File: 01-Programming-Languages/Ruby/Data-Structures/LinkedList/reverse_linked_list.rb
=begin
Problem: Reverse Linked List
Description: Reverse a singly linked list in-place.

Approach:
- Use three pointers: prev, curr, next
- Point current node's next to prev, then advance

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input:  1 -> 2 -> 3 -> 4 -> 5
Output: 5 -> 4 -> 3 -> 2 -> 1
=end

class ListNode
  attr_accessor :val, :next

  def initialize(val = 0, nxt = nil)
    @val = val
    @next = nxt
  end
end

def reverse_list(head)
  prev = nil
  curr = head
  while curr
    nxt = curr.next
    curr.next = prev
    prev = curr
    curr = nxt
  end
  prev
end

def print_list(head)
  vals = []
  while head
    vals << head.val
    head = head.next
  end
  puts vals.join(' -> ')
end

head = ListNode.new(1, ListNode.new(2, ListNode.new(3, ListNode.new(4, ListNode.new(5)))))
puts "Original: "
print_list(head)
reversed = reverse_list(head)
puts "Reversed: "
print_list(reversed)
